pkg/cloudprovider/vsphereparavirtual/route.go

```diff
@@ -0,0 +1,415 @@
+		routeSetClusterName := routeSet.Annotations[AnnotationClusterName]
+		if routeSetClusterName == clusterName {
+			// only return cloudprovider.Route if RouteSet CR status 'Ready' is true
+			// condition := GetRouteSetCondition(&(routeSet.Status), v1alpha1.RouteSetConditionTypeReady)
+			// if condition != nil && condition.Status == v1.ConditionTrue {
+			// one RouteSet per node, so we can use nodeName as the name of RouteSet CR
+			nodeName := routeSet.Name
+			for _, route := range routeSet.Spec.Routes {
+				cpRoute := &cloudprovider.Route{
+					Name:            route.Name,
+					TargetNode:      types.NodeName(nodeName),
+					DestinationCIDR: route.Destination,
+				}
+				routes = append(routes, cpRoute)
+			}
+			// }
+		}
+	}
+	return routes
+}
+
+// CreateRoute implements Routes.CreateRoute
+// Create a RouteSet custom resource for a Node
+func (r *routesProvider) CreateRoute(ctx context.Context, clusterName string, nameHint string, route *cloudprovider.Route) error {
+	nodeName := string(route.TargetNode)
+	klog.V(6).Infof("Creating Route for node %s with hint %s in cluster %s", nodeName, nameHint, clusterName)
+
+	nodeIP, err := r.getNodeIPAddress(nodeName, util.IsIPv4(route.DestinationCIDR))
+	if err != nil {
+		klog.Errorf("getting node %s IP address failed: %v", nodeName, err)
+		return err
+	}
+
+	_, err = r.createRouteSetCR(ctx, clusterName, nameHint, nodeName, route.DestinationCIDR, nodeIP)
+	if err != nil {
+		klog.Errorf("creating RouteSet CR for node %s failed: %s", nodeName, err)
+		return err
+	}
+	return nil
+	// TODO: Get realized state of corresponding static route
+	// return r.checkStaticRouteRealizedState(routeSet.Name)
+}
+
+// createRouteSetCR creates RouteSet CR through RouteSet client
+func (r *routesProvider) createRouteSetCR(ctx context.Context, clusterName string, nameHint string, nodeName string, cidr string, nodeIP string) (*v1alpha1.RouteSet, error) {
+	annotations := map[string]string{
+		AnnotationClusterName: clusterName,
+	}
+	nodeRef := metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       nodeName,
+		UID:        types.UID(nameHint),
+	}
+	r.ownerRefs = append(r.ownerRefs, nodeRef)
+	route := v1alpha1.Route{
+		Name:        r.GetRouteName(nodeName, cidr, clusterName),
+		Destination: cidr,
+		Target:      nodeIP,
+	}
+	routeSetSpec := v1alpha1.RouteSetSpec{
+		Routes: []v1alpha1.Route{
+			route,
+		},
+	}
+	routeSet := &v1alpha1.RouteSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations:     annotations,
+			Name:            nodeName,
+			OwnerReferences: r.ownerRefs,
+			Namespace:       r.namespace,
+		},
+		Spec: routeSetSpec,
+	}
+
+	_, err := r.routeClient.NsxV1alpha1().RouteSets(r.namespace).Create(ctx, routeSet, metav1.CreateOptions{})
+	if err != nil {
+		klog.ErrorS(ErrCreateRouteSet, fmt.Sprintf("%v", err))
+		return nil, err
+	}
+
+	klog.V(6).Infof("Successfully created RouteSet CR for node %s", nodeName)
+	err = r.updateNetworkingCondition(nodeName)
```

Comment:
Is it necessary to update `NodeNetworkUnavailable` here?
If I remember it correctly, route controller will update it after it gets `ListRoutes` back?
If route controller does not do that, I think it's better to update `NodeNetworkUnavailable` only for realized routes.

Reply:
Yeah the cloud-provider does add this condition

https://github.com/kubernetes/cloud-provider/blob/master/controllers/route/route_controller.go#L270

Reply:
https://github.com/kubernetes/cloud-provider/blob/master/controllers/route/route_controller.go#L111-L115
The upstream is doing the resync every 5 minutes. maybe that's too long, but still it should be fixed upstream

Reply:
yeah I agree, this is something we need to fix in the upstream - 5 mins is too long
sometimes I see cloud provider Pod pending because of this, that's why I added the update here
but I agree that it might not be good to add `updateNetworkingCondition` here, let me remove this for now and we can fix it in the upstream later

Reply:
https://github.com/kubernetes/cloud-provider/blob/5a236942a3ca6f0454bfcb979cc31207af4894b2/config/v1alpha1/defaults.go#L66
The comment about 5 minutes may be wrong, it should be 10 seconds. So I think removing updateNetworkingCondition here should be fine.